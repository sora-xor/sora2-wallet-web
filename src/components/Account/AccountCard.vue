<template>
  <s-card v-bind="{ shadow: 'always', size: 'small', borderRadius: 'medium', ...$attrs }" class="account-card">
    <div class="account">
      <div class="account-avatar">
        <slot name="avatar" />
      </div>
      <div class="account-details s-flex">
        <div class="account-credentials s-flex">
          <div class="account-credentials_name">
            <slot name="name" />
          </div>
          <div class="account-credentials_description">
            <slot name="description" />
          </div>
        </div>
        <slot />
      </div>
    </div>
  </s-card>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

@Component
export default class AccountCard extends Vue {}
</script>

<style lang="scss">
.account-card {
  display: flex;
  align-items: center;

  & > .el-card__body {
    flex: 1;
    max-width: 100%;
  }

  &.s-card.neumorphic {
    border-width: 1px;
    border-style: solid;
    border-color: transparent;
  }
}

.account {
  &-avatar {
    & > img {
      max-width: 100%;
    }
  }
  &-details {
    .el-button + .el-button {
      margin-left: 0;
    }
  }
}
</style>

<style scoped lang="scss">
$gap: $basic-spacing-small;
$avatar-size: 32px;

.account {
  display: flex;
  align-items: center;
  gap: $gap;

  &-avatar {
    display: flex;
    align-items: center;
    flex-shrink: 0;
    width: $avatar-size;
    height: $avatar-size;
    border-radius: 50%;
    overflow: hidden;
  }

  &-details {
    flex: 1;
    align-items: center;
    max-width: calc(100% - $gap - $avatar-size);
  }

  &-credentials {
    flex: 1;
    flex-direction: column;
    justify-content: center;
    overflow: hidden;

    &_name,
    &_description {
      overflow: hidden;
      text-overflow: ellipsis;
      letter-spacing: var(--s-letter-spacing-small);
    }

    &_name {
      font-size: var(--s-font-size-medium);
      font-weight: 600;
      line-height: var(--s-line-height-medium);
      outline: none;
      white-space: nowrap;
      text-overflow: ellipsis;
    }

    &_description {
      @include value-prefix(width, fit-content);
      @include hint-text;
      outline: none;
    }
  }
}
</style>
