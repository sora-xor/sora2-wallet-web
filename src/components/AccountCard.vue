<template>
  <s-card v-bind="{ shadow: 'always', size: 'small', borderRadius: 'medium', ...$attrs }" class="wallet-account">
    <div class="account s-flex">
      <div class="account-avatar">
        <slot name="avatar" />
      </div>
      <div class="account-details s-flex">
        <div class="account-credentials s-flex">
          <div class="account-credentials_name">
            <slot name="name" />
          </div>
          <div class="account-credentials_description">
            <slot name="description" />
          </div>
        </div>
        <slot />
      </div>
    </div>
  </s-card>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

@Component
export default class AccountCard extends Vue {}
</script>

<style lang="scss">
.account {
  &-details {
    .el-button + .el-button {
      margin-left: 0;
    }
  }
}
</style>

<style scoped lang="scss">
$avatar-margin-right: #{$basic-spacing-small};
$avatar-size: 32px;

.wallet-account.s-card {
  border: 1px solid transparent;
}

.account {
  align-items: center;

  &-avatar {
    display: flex;
    flex-shrink: 0;
    margin-right: $avatar-margin-right;
    width: $avatar-size;
    height: $avatar-size;
  }

  &-details {
    flex: 1;
    align-items: center;
    width: calc(100% - #{$avatar-size} - #{$avatar-margin-right});
  }

  &-credentials {
    flex: 1;
    flex-direction: column;
    justify-content: center;
    overflow: hidden;

    &_name,
    &_description {
      overflow: hidden;
      text-overflow: ellipsis;
      letter-spacing: var(--s-letter-spacing-small);
    }

    &_name {
      font-size: var(--s-font-size-medium);
      font-weight: 600;
      line-height: var(--s-line-height-medium);
      outline: none;
      white-space: nowrap;
      text-overflow: ellipsis;
    }

    &_description {
      @include value-prefix(width, fit-content);
      @include hint-text;
      outline: none;
    }
  }
}
</style>
