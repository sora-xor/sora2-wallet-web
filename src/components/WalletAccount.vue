<template>
  <s-card shadow="always" size="small" border-radius="medium" class="wallet-account">
    <div class="account s-flex">
      <wallet-avatar class="account-avatar" :address="address" :size="28" />
      <div class="account-details s-flex">
        <div class="account-credentials s-flex">
          <div v-if="name" class="account-credentials_name">{{ name }}</div>
          <s-tooltip :content="t('account.copy')">
            <div class="account-credentials_address" @click="handleCopyAddress($event)">{{ formattedAddress }}</div>
          </s-tooltip>
        </div>
        <template v-if="showControls">
          <s-button
            class="account__action-button account-switch"
            type="action"
            alternative
            rounded
            :tooltip="t('account.switch')"
            @click="handleSwitchAccount"
          >
            <s-icon name="arrows-refresh-ccw-24" size="28" />
          </s-button>
          <s-button
            class="account__action-button account-logout"
            type="action"
            alternative
            rounded
            :tooltip="t('account.logout')"
            @click="handleLogout"
          >
            <s-icon name="security-logout-24" size="28" />
          </s-button>
        </template>
      </div>
    </div>
  </s-card>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator'
import { Getter, Action } from 'vuex-class'

import TranslationMixin from './mixins/TranslationMixin'
import { RouteNames } from '../consts'
import { copyToClipboard, formatAddress, formatSoraAddress } from '../util'
import WalletAvatar from './WalletAvatar.vue'

@Component({
  components: {
    WalletAvatar
  }
})
export default class WalletAccount extends Mixins(TranslationMixin) {
  @Getter account!: any
  @Action logout
  @Action navigate

  @Prop({ default: false, type: Boolean }) readonly showControls!: boolean
  @Prop({ default: () => null, type: Object }) readonly polkadotAccount!: { name: string; address: string }

  get address (): string {
    if (this.polkadotAccount) {
      return formatSoraAddress(this.polkadotAccount.address)
    }
    return this.account.address
  }

  get name (): string {
    return (this.polkadotAccount || this.account).name
  }

  get formattedAddress (): string {
    return formatAddress(this.address, 24)
  }

  async handleCopyAddress (event: Event): Promise<void> {
    event.stopImmediatePropagation()
    try {
      await copyToClipboard(this.address)
      this.$notify({
        message: this.t('account.successCopy'),
        type: 'success',
        title: ''
      })
    } catch (error) {
      this.$notify({
        message: `${this.t('warningText')} ${error}`,
        type: 'warning',
        title: ''
      })
    }
  }

  private navigateToWalletConnection (params?: any): void {
    const navigationArgs: any = { name: RouteNames.WalletConnection }
    if (params) {
      navigationArgs.params = params
    }
    this.navigate(navigationArgs)
    this.logout()
  }

  handleSwitchAccount (): void {
    this.navigateToWalletConnection({ isAccountSwitch: true })
  }

  handleLogout (): void {
    this.navigateToWalletConnection()
  }
}
</script>

<style lang="scss">
.account-avatar {
  border: 2px solid var(--s-color-base-border-secondary);
  border-radius: 50%;
  svg circle:first-child {
    fill: var(--s-color-utility-surface);
  }
}
</style>

<style scoped lang="scss">
@import '../styles/icons';

$avatar-margin-right: #{$basic-spacing-small};
$avatar-size: 32px;

.wallet-account.s-card {
  border: 1px solid transparent;
}

.account {
  align-items: center;

  &-avatar {
    margin-right: $avatar-margin-right;
    width: $avatar-size;
    height: $avatar-size;
    flex-shrink: 0;
  }
  &-details {
    flex: 1;
    width: calc(100% - #{$avatar-size} - #{$avatar-margin-right});
  }
  &-credentials {
    flex: 1;
    flex-direction: column;
    justify-content: center;
    &_name,
    &_address {
      overflow: hidden;
      text-overflow: ellipsis;
      letter-spacing: var(--s-letter-spacing-small);
    }
    &_name {
      font-size: var(--s-font-size-medium);
      font-weight: 600;
      line-height: var(--s-line-height-medium);
      outline: none;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }
    &_address {
      @include value-prefix(width, fit-content);
      @include hint-text;
      outline: none;
      &:hover {
        text-decoration: underline;
        cursor: pointer;
      }
    }
  }
  &__action-button {
    & + & {
      margin-left: var(--s-basic-spacing);
    }
  }
  &-switch, &-logout {
    padding: 0;
  }
}
</style>

<style lang="scss">
.account-menu i:hover {
  color: var(--s-color-theme-accent);
}
</style>
