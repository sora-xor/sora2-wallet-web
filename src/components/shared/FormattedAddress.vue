<template>
  <s-tooltip :content="copyTooltip(tooltipText)" tabindex="-1" append-to-body>
    <div class="formatted-address" @click="handleCopyAddress(value, $event)">
      <template v-if="sliced">
        <span class="address" :style="{ width: firstPartWidth }">{{ value }}</span>
        ...
        <span>{{ secondPart }}</span>
      </template>
      <template v-else>
        {{ value }}
      </template>
    </div>
  </s-tooltip>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator';

import { getTextWidth } from '../../util';
import CopyAddressMixin from '../mixins/CopyAddressMixin';

@Component
export default class FormattedAddress extends Mixins(CopyAddressMixin) {
  @Prop({ default: '', type: String }) readonly value!: string;
  @Prop({ default: '', type: String }) readonly tooltipText!: string;
  /** Default visible address length, default: 12 */
  @Prop({ default: 12, type: [Number, String] }) readonly symbols!: number | string;
  /** Offset in px, default: 0 */
  @Prop({ default: 0, type: [Number, String] }) readonly offset!: number | string;
  /** Offset in symbols, default: 0 */
  @Prop({ default: 0, type: [Number, String] }) readonly symbolsOffset!: number | string;

  get sliced(): boolean {
    return this.value.length >= +this.symbols;
  }

  get count(): number {
    return +this.symbols / 2;
  }

  get firstPartWidth(): string {
    const text = this.value.slice(0, Math.round(this.count) + Number(this.symbolsOffset));
    const width = getTextWidth(text) - Number(this.offset);
    return width + 'px';
  }

  get secondPart(): string {
    // TODO: Remove this dirty hack. It's made just for browser search
    return this.value.slice(-Math.floor(this.count) + Number(this.symbolsOffset));
  }
}
</script>

<style lang="scss" scoped>
.formatted-address {
  @include hint-text;

  display: inline-flex;
  color: inherit;
  cursor: pointer;
  letter-spacing: normal;
  white-space: nowrap;

  &:hover {
    text-decoration: underline;
  }

  .address {
    white-space: nowrap;
    overflow: hidden;
  }
}
</style>
