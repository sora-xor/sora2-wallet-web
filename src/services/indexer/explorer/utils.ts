import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
