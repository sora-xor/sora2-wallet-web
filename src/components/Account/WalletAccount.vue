<template>
  <account-card v-bind="$attrs">
    <template #avatar>
      <wallet-avatar slot="avatar" class="account-gravatar" :address="address" :size="28" />
    </template>
    <template #name>{{ name }}</template>
    <template #description>
      <!-- TODO: Add Copy address by Keyboard -->
      <s-tooltip :content="copyTooltip(t('account.walletAddress'))" tabindex="-1">
        <div class="account-credentials_address" @click="handleCopyAddress(address, $event)">
          <p class="first">{{ address }}</p>
          ...
          <p>{{ secondPartOfAddress }}</p>
        </div>
      </s-tooltip>
    </template>
    <template #default>
      <slot />
    </template>
  </account-card>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator';

import { ObjectInit } from '../../consts';
import { getter } from '../../store/decorators';
import { formatAddress, formatSoraAddress } from '../../util';
import CopyAddressMixin from '../mixins/CopyAddressMixin';
import TranslationMixin from '../mixins/TranslationMixin';
import WalletAvatar from '../WalletAvatar.vue';

import AccountCard from './AccountCard.vue';

import type { PolkadotJsAccount } from '../../types/common';

const ADDRESS_LENGTH = 24;

@Component({
  components: {
    AccountCard,
    WalletAvatar,
  },
})
export default class WalletAccount extends Mixins(TranslationMixin, CopyAddressMixin) {
  @getter.account.account private account!: PolkadotJsAccount;

  @Prop({ default: ObjectInit, type: Object }) readonly polkadotAccount!: PolkadotJsAccount;

  get address(): string {
    if (this.polkadotAccount) {
      return formatSoraAddress(this.polkadotAccount.address);
    }
    return this.account.address;
  }

  get name(): string {
    return (this.polkadotAccount || this.account).name;
  }

  get formattedAddress(): string {
    return formatAddress(this.address, ADDRESS_LENGTH);
  }

  get secondPartOfAddress(): string {
    // TODO: Remove this dirty hack. It's made just for browser search
    return this.address.slice(-ADDRESS_LENGTH / 2);
  }
}
</script>

<style lang="scss">
.account-gravatar {
  border: 2px solid var(--s-color-base-border-secondary);
  border-radius: 50%;

  & > circle:first-child {
    fill: var(--s-color-utility-surface);
  }
}
</style>

<style scoped lang="scss">
.account-credentials_address {
  display: flex;
  .first {
    width: 100px;
    white-space: nowrap;
    overflow: hidden;
  }
  &:hover {
    text-decoration: underline;
    cursor: pointer;
  }
}
</style>
