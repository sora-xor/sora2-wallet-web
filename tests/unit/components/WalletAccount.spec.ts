import AccountCard from '@/components/Account/AccountCard.vue';
import WalletAccount from '@/components/Account/WalletAccount.vue';
import FormattedAddress from '@/components/shared/FormattedAddress.vue';

import { useDescribe, useShallowMount, useVuex } from '../../utils';
import { MOCK_ACCOUNT, MOCK_ACCOUNT_POLKADOT, MOCK_ADDRESS } from '../../utils/WalletAccountMock';

const createStore = () =>
  useVuex({
    account: {
      state: () => ({
        name: MOCK_ACCOUNT.name,
        address: MOCK_ACCOUNT.address,
        source: MOCK_ACCOUNT.source,
      }),
      getters: {
        account: () => MOCK_ACCOUNT,
      },
    },
  });

const mockFormattedSoraAddress = jest.fn().mockReturnValue(MOCK_ADDRESS.formattedSora);

jest.mock('../../../src/util', () => {
  const originalModule = jest.requireActual('../../../src/util');

  return {
    ...originalModule,
    formatSoraAddress: () => mockFormattedSoraAddress(),
  };
});

useDescribe('WalletAccount.vue', WalletAccount, () => {
  let wrapper;
  const polkadotAccount = MOCK_ACCOUNT_POLKADOT;
  const wrapperOptions = {
    propsData: {
      polkadotAccount,
    },
    store: createStore(),
    stubs: {
      AccountCard,
      FormattedAddress,
    },
  };

  beforeEach(() => {
    wrapper = useShallowMount(WalletAccount, wrapperOptions);
  });

  it('should be rendered correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('account should have name', () => {
    const div = wrapper.find('.account-credentials_name');

    expect(div.text()).toMatch(MOCK_ACCOUNT_POLKADOT.name);
  });

  // [TODO] Fix it
  // it('account should have formatted address', () => {
  //   const div = wrapper.find('.account-credentials_address');

  //   expect(div.text()).toBe('cnRXua6zs8TaE87BQFL6uWVbT2g6GXsUjwk6PTvL6UHcHDCvo...TvL6UHcHDCvo');
  // });

  it('should use general account name and address', () => {
    const wrapper = useShallowMount(WalletAccount, {
      ...wrapperOptions,
      propsData: {
        polkadotAccount: null,
      },
    });

    const divName = wrapper.find('.account-credentials_name');
    const divAddressText = wrapper.find('.formatted-address .address').text();
    const startLine = divAddressText.substring(0, 12);
    const endLine = divAddressText.substring(15);
    const accountGetter = wrapper.vm.$store.getters['wallet/account/account'];

    expect(accountGetter.name).toBe(divName.text());
    expect(accountGetter.address).toStartWith(startLine);
    // [TODO]: Fix it
    expect(accountGetter.address).toEndWith('BQFL6uWVbT2g6GXsUjwk6PTvL6UHcHDCvo');
  });
});
