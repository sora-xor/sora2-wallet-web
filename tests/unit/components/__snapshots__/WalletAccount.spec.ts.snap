// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletAccount.vue should be rendered correctly 1`] = `
<s-card
  borderradius="medium"
  class="account-card"
  shadow="always"
  size="small"
>
  <div
    class="account"
  >
    <div
      class="account-avatar"
    >
      <wallet-avatar-stub
        address="cnRXua6zs8TaE87BQFL6uWVbT2g6GXsUjwk6PTvL6UHcHDCvo"
        class="account-gravatar"
        size="28"
        slot="avatar"
        theme="polkadot"
      />
    </div>
     
    <div
      class="account-details s-flex"
    >
      <div
        class="account-credentials s-flex"
      >
        <div
          class="account-credentials_name"
        >
          Polkadot Name
        </div>
         
        <div
          class="account-credentials_description"
        >
          <s-tooltip-stub
            animation="el-fade-in-linear"
            borderradius="small"
            closedelay="0"
            content="Copy Wallet address"
            hideafter="0"
            offset="0"
            opendelay="500"
            placement="top"
            showarrow="true"
            tabindex="-1"
            theme="auto"
            usedesignsystem="true"
          >
            <div
              class="account-credentials_address"
            >
              <p
                class="first"
              >
                cnRXua6zs8TaE87BQFL6uWVbT2g6GXsUjwk6PTvL6UHcHDCvo
              </p>
              
        ...
        
              <p>
                TvL6UHcHDCvo
              </p>
            </div>
          </s-tooltip-stub>
        </div>
      </div>
       
    </div>
  </div>
</s-card>
`;
